Use 'vanilla' websocket components

guusdk opened the pull request:
**Don't merge this PR without a thorough review!**

This commit attempts to make our lives easier by removing the custom websocket implementation in Openfire, Strophe, and the clients that make use of Strophe.

Fewer customizations gives us:
- less problems upgrading third party components (I got a couple of other idea's too - I'll run them by you later).
- less complexity in *our* code. Helps to maintain stuff, especially when one is not the original author of everything. ;)

This does introduce a dependency on Openfire's Websocket plugin - but I think that's a good trade-off (perhaps something for the future: do we *need* websockets, or can we use BOSH and optionally use websockets when available?)

I am aware that this potentially touches *a lot* - hence my warning above. 

Tested so far:
- Jitsi Meet
- Candy

Untested:
- fastpath
- chrome extension
- remote control
- ... (anything I'm unaware of)

deleolajide commented:
My first question without going through the code changes is how are we now implementing single authentication for webservice and XMPP connection from client?

I would like to prevent user having to enter username/password twice. The current implementation enables the web socket in strophe to use the cached browser authentication. There is a check server-side for the existence of the user authentication in the web service and reused for the websocket. As both are implemented in Jetty, this integration is simple with lots of benefit to the user. No need for XMPP user authentication speeding up the connection time.

We have to address this issue and provide such an integration from the standard openfire websocket in order not to lose this feature.

guusdk commented:
I thought I had this covered, as I didn't get a second login popup. Turns out that Jitsi Meet does not give you one that visibly. I am indeed not logged in with the XMPP client. Back to the drawing board...

(I did rebase the PR, to account for the other merge)

guusdk commented:
My last commit should resolve the authentication issue that you raised. As we discussed out-of-band, I've replaced the old implementation with one that lets Jetty register an access token for a user. The web applications then use that token to authenticate.

To authenticate, I've added a new SASL mechanism (which is the preferred authentication solution in XMPP anyway). Through the OFMeet plugin, I've added support to Openfire for that mechanism. Client support has been added by a new plugin for Strophe (which is used by Jitsi Meet as well as by Candy).

deleolajide commented:
Guus, Thanks for doing this. I have however made this change a little more complicated by adding support for Azure AD for users using usernames of type user@domain on release 0.3.20. In your code, you are throwing an exception for this situation.

I will proceed to pull these change request and re-apply the Azure AD support code after I get my head around your changes.

My only misgiving about this change is that default connection behaviour of Openfire Meetings will now be BOSH and Websockets becomes an optional feature.

guusdk commented:
Dele, fire up your webchat. I left you a bunch of messages on the Azure
subject.

On 3 May 2016 at 16:15, Dele Olajide <notifications@github.com> wrote:

> Guus, Thanks for doing this. I have however made this change a little more
> complicated by adding support for Azure AD for users using usernames of
> type user@domain on release 0.3.20. In your code, you are throwing an
> exception for this situation.
>
> I will proceed to pull these change request and re-apply the Azure AD
> support code after I get my head around your changes.
>
> My only misgiving about this change is that default connection behaviour
> of Openfire Meetings will now be BOSH and Websockets becomes an optional
> feature.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/igniterealtime/community-plugins/pull/19#issuecomment-216541883>
>
